File: mobile-app/src/services/authService.ts
import { storage } from './storage';
import api from './api';
import { AuthResponse, User } from '../types';

const TOKEN_KEY = 'auth_token';
const REFRESH_KEY = 'refresh_token';
const USER_KEY = 'user_data';

interface ApiResponse<T> {
  success: boolean;
  data: T;
  message?: string;
}

export const authService = {
  async login(email: string, password: string): Promise<AuthResponse> {
    const response = await api.post<ApiResponse<AuthResponse>>('/auth/login', {
      email: email.toLowerCase().trim(),
      password,
    });

    const { accessToken, refreshToken, user } = response.data.data;

    await storage.setItem(TOKEN_KEY, accessToken);
    if (refreshToken) {
      await storage.setItem(REFRESH_KEY, refreshToken);
    }
    await storage.setItem(USER_KEY, JSON.stringify(user));

    return { accessToken, refreshToken, user };
  },

  async logout(): Promise<void> {
    try {
      await api.post('/auth/logout');
    } catch {
      // ignore logout errors
    }
    await storage.removeItem(TOKEN_KEY);
    await storage.removeItem(REFRESH_KEY);
    await storage.removeItem(USER_KEY);
  },

  async refreshToken(): Promise<string> {
    const refreshToken = await storage.getItem(REFRESH_KEY);
    if (!refreshToken) throw new Error('No refresh token available');

    const response = await api.post<ApiResponse<{ accessToken: string; refreshToken?: string }>>(
      '/auth/refresh',
      { refreshToken }
    );

    const { accessToken, refreshToken: newRefresh } = response.data.data;
    await storage.setItem(TOKEN_KEY, accessToken);
    if (newRefresh) {
      await storage.setItem(REFRESH_KEY, newRefresh);
    }

    return accessToken;
  },

  async getProfile(): Promise<User> {
    const response = await api.get<ApiResponse<User>>('/auth/me');
    return response.data.data;
  },

  async changePassword(
    currentPassword: string,
    newPassword: string
  ): Promise<void> {
    await api.put('/auth/change-password', {
      currentPassword,
      newPassword,
    });
  },

  async getStoredUser(): Promise<User | null> {
    try {
      const data = await storage.getItem(USER_KEY);
      return data ? JSON.parse(data) : null;
    } catch {
      return null;
    }
  },

  async getStoredToken(): Promise<string | null> {
    try {
      return await storage.getItem(TOKEN_KEY);
    } catch {
      return null;
    }
  },
};

export default authService;

File: mobile-app/lib/services/auth_service.dart
import 'dart:convert';
import '../models/user.dart';
import '../models/api_response.dart';
import 'api_client.dart';
import 'api_endpoints.dart';

class AuthService {
  final ApiClient _api = ApiClient();

  Future<ApiResponse<LoginResponse>> login(String email, String password) async {
    try {
      final res = await _api.post(ApiEndpoints.login, body: {
        'email': email,
        'password': password,
      }, auth: false);

      final body = jsonDecode(res.body) as Map<String, dynamic>;

      if (res.statusCode == 200 && body['success'] == true) {
        final loginRes = LoginResponse.fromJson(body['data'] as Map<String, dynamic>);
        await _api.saveTokens(loginRes.accessToken, loginRes.refreshToken);
        await _api.saveCredentials(email, password);
        return ApiResponse(success: true, data: loginRes);
      }

      final rawMessage = body['message'] as String?;
      final rawErrors = body['errors'];
      String message;
      if (rawErrors is Map) {
        message = (rawErrors as Map<String, dynamic>).entries
            .map((e) => (e.value as List).join('; '))
            .join('\n');
      } else if (rawMessage != null && rawMessage != 'Validation failed') {
        message = rawMessage;
      } else {
        message = 'Invalid email or password';
      }
      return ApiResponse(
        success: false,
        message: message,
        error: body['error'] as String?,
      );
    } catch (e) {
      return ApiResponse(success: false, message: 'Network error. Please try again.');
    }
  }

  Future<ApiResponse<User>> getProfile() async {
    try {
      final res = await _api.get(ApiEndpoints.me);
      final body = jsonDecode(res.body) as Map<String, dynamic>;

      if (res.statusCode == 200 && body['success'] == true) {
        final user = User.fromJson(body['data'] as Map<String, dynamic>);
        return ApiResponse(success: true, data: user);
      }

      return ApiResponse(success: false, message: body['message'] as String?);
    } catch (e) {
      return ApiResponse(success: false, message: 'Failed to load profile');
    }
  }

  Future<void> logout() async {
    await _api.clearCredentials();
    await _api.clearTokens();
  }

  Future<bool> isLoggedIn() async {
    return _api.isLoggedIn;
  }
}
